File: web/src/protocol.ts
/**
 * The wire contract between the TypeScript front end and the Rust backend.
 *
 * This file is hand-written for now. Once the Rust side exists, it should be
 * REPLACED by output from `ts-rs` (`#[derive(TS)]` on the Rust types, generated
 * during `cargo test`) so that drift between the two fails CI rather than
 * showing up as a blank order book in the pub.
 *
 * Conventions:
 *  - Every field is camelCase on the wire. On the Rust side that means
 *    `#[serde(rename_all = "camelCase")]` on each struct.
 *  - Enums are internally tagged on `t`, i.e. `#[serde(tag = "t", rename_all = "camelCase")]`.
 *  - `price` is a float at the UI edge. The engine stores scaled i64 micro-units
 *    internally; conversion happens at the serde boundary, not here.
 *  - Timestamps are epoch milliseconds.
 *
 * Note on sizes: every order is one lot, so a trade always fully consumes exactly
 * one resting order. There are no partial fills anywhere in this protocol.
 */

export type Phase = "lobby" | "open" | "closed" | "settled";

/** Which side of the book an order rests on. */
export type Side = "bid" | "offer";

/** What an aggressor did. MINE => "buy", YOURS => "sell". */
export type Direction = "buy" | "sell";

export interface SessionMeta {
  code: string;
  question: string;
  /** Host-defined unit label, e.g. "millions of balls". Display only. */
  unit: string;
  /** null means any price is accepted. */
  tickSize: number | null;
  positionLimit: number;
  phase: Phase;
}

export interface Player {
  id: string;
  name: string;
  joinedAt: number;
}

export interface Order {
  id: string;
  playerId: string;
  /** Denormalised so the projector can render names without a player lookup. */
  playerName: string;
  side: Side;
  price: number;
  /** The sequence number at which this order entered the book. This IS its time
   *  priority — the engine must not rely on wall-clock time for ordering. */
  seq: number;
}

export interface Trade {
  id: string;
  seq: number;
  price: number;
  buyerId: string;
  buyerName: string;
  sellerId: string;
  sellerName: string;
  /** Which side crossed the spread. */
  aggressor: Direction;
  /** True when buyer and seller are the same player. Allowed by design — see
   *  DECISIONS.md #19 — but tagged so the tape does not look like a bug. */
  selfTrade: boolean;
  ts: number;
}

export interface BookState {
  /** Best first: highest price, then lowest seq. */
  bids: Order[];
  /** Best first: lowest price, then lowest seq. */
  offers: Order[];
}

/** Per-connection private state. Never contains another player's position. */
export interface YouState {
  playerId: string;
  name: string;
  position: number;
}

export interface Result {
  playerId: string;
  name: string;
  position: number;
  /** Sum of sell prices minus sum of buy prices. */
  cash: number;
  /** cash + position * trueValue */
  pnl: number;
}

export type RejectReason =
  | "position_limit"
  | "not_open"
  | "no_liquidity"
  | "bad_tick"
  | "bad_price"
  | "unknown_order"
  | "not_your_order"
  | "not_host"
  | "unknown_player";

/* ------------------------------------------------------------------ */
/* Server -> client                                                    */
/* ------------------------------------------------------------------ */

/**
 * Every event except `rejected` and `error` carries a monotonic `seq` and is
 * broadcast to every connection in the session. A client that sees a gap in
 * `seq` must send `resync` and will receive a fresh `snapshot`.
 *
 * `rejected` and `error` are addressed to a single connection and carry no seq,
 * because they are not part of the shared event log.
 */
export type ServerEvent =
  | {
      t: "snapshot";
      seq: number;
      session: SessionMeta;
      players: Player[];
      book: BookState;
      /** Full tape, oldest first. */
      trades: Trade[];
      /** null for a host connection — the host does not trade. */
      you: YouState | null;
      /** Present only once phase is "settled". */
      settlement: { trueValue: number; results: Result[] } | null;
    }
  | { t: "playerJoined"; seq: number; player: Player }
  | { t: "phaseChanged"; seq: number; phase: Phase }
  | { t: "orderAdded"; seq: number; order: Order }
  | { t: "orderCancelled"; seq: number; orderId: string; playerId: string }
  | {
      t: "trade";
      seq: number;
      trade: Trade;
      /** The resting order consumed by this trade. Clients remove it from the
       *  book. Always fully consumed, since every order is one lot. */
      restingOrderId: string;
    }
  | { t: "settled"; seq: number; trueValue: number; results: Result[] }
  | { t: "rejected"; reason: RejectReason; message: string }
  | { t: "error"; message: string };

/* ------------------------------------------------------------------ */
/* Client -> server                                                    */
/* ------------------------------------------------------------------ */

export type ClientCommand =
  /** Rest a new order. Players may hold any number per side. */
  | { t: "placeOrder"; side: Side; price: number }
  | { t: "cancelOrder"; orderId: string }
  /** MINE (`buy`) lifts the best offer. YOURS (`sell`) hits the best bid.
   *  Rejected with `no_liquidity` if that side of the book is empty. */
  | { t: "take"; direction: Direction }
  /** Sent after a detected seq gap, or on reconnect. */
  | { t: "resync"; fromSeq: number }
  /* Host-only. Rejected with `not_host` from a player connection. */
  | { t: "openTrading" }
  | { t: "closeTrading" }
  | { t: "settle"; trueValue: number };

/* ------------------------------------------------------------------ */
/* HTTP payloads                                                       */
/* ------------------------------------------------------------------ */

export interface CreateSessionRequest {
  question: string;
  unit: string;
  tickSize: number | null;
  positionLimit: number;
}

export interface CreateSessionResponse {
  code: string;
  /** Secret. Stored in the host's localStorage; gates open/close/settle. */
  hostToken: string;
}

export interface JoinSessionRequest {
  name: string;
}

export interface JoinSessionResponse {
  playerId: string;
  /** Secret. Stored in the player's localStorage so a screen-lock or refresh
   *  restores the same seat rather than creating a new trader. */
  playerToken: string;
}

/* ------------------------------------------------------------------ */
/* Helpers                                                             */
/* ------------------------------------------------------------------ */

export const bestBid = (book: BookState): Order | undefined => book.bids[0];
export const bestOffer = (book: BookState): Order | undefined => book.offers[0];

export function spread(book: BookState): number | null {
  const b = bestBid(book);
  const o = bestOffer(book);
  return b && o ? o.price - b.price : null;
}

/** Position is derivable from the public tape, since trades carry names. */
export function positionFrom(trades: Trade[], playerId: string): number {
  let pos = 0;
  for (const tr of trades) {
    if (tr.buyerId === playerId) pos += 1;
    if (tr.sellerId === playerId) pos -= 1;
  }
  return pos;
}

export function tradesInvolving(trades: Trade[], playerId: string): Trade[] {
  return trades.filter((t) => t.buyerId === playerId || t.sellerId === playerId);
}

export function formatPrice(price: number, tickSize: number | null): string {
  if (tickSize && tickSize < 1) {
    const dp = Math.min(6, (String(tickSize).split(".")[1] ?? "").length);
    return price.toFixed(dp);
  }
  return Number.isInteger(price) ? String(price) : String(Number(price.toFixed(4)));
}

File: server/src/protocol.rs
//! Wire types. These mirror `web/src/protocol.ts` exactly.
//!
//! Right now the two are kept in step by hand, which will not last. The fix is
//! `ts-rs`: derive `TS` on everything here and let `cargo test` generate the
//! TypeScript, so a mismatch breaks the build rather than the game.

use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub enum Phase {
    Lobby,
    Open,
    Closed,
    Settled,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub enum Side {
    Bid,
    Offer,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub enum Direction {
    Buy,
    Sell,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub struct SessionMeta {
    pub code: String,
    pub question: String,
    pub unit: String,
    pub tick_size: Option<f64>,
    #[ts(type = "number")]
    pub position_limit: i64,
    pub phase: Phase,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub struct Player {
    pub id: String,
    pub name: String,
    #[ts(type = "number")]
    pub joined_at: i64,
    /// Bots are ordinary players to the engine — they hold positions, respect
    /// the limit and settle like anyone else. This only tells the UI to mark
    /// them so the room knows who it is trading against.
    pub is_bot: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub struct Order {
    pub id: String,
    pub player_id: String,
    pub player_name: String,
    pub side: Side,
    pub price: f64,
    #[ts(type = "number")]
    pub seq: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub struct Trade {
    pub id: String,
    #[ts(type = "number")]
    pub seq: u64,
    pub price: f64,
    pub buyer_id: String,
    pub buyer_name: String,
    pub seller_id: String,
    pub seller_name: String,
    pub aggressor: Direction,
    pub self_trade: bool,
    #[ts(type = "number")]
    pub ts: i64,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub struct BookState {
    pub bids: Vec<Order>,
    pub offers: Vec<Order>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub struct YouState {
    pub player_id: String,
    pub name: String,
    #[ts(type = "number")]
    pub position: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub struct Result {
    pub player_id: String,
    pub name: String,
    #[ts(type = "number")]
    pub position: i64,
    pub cash: f64,
    pub pnl: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub struct Settlement {
    pub true_value: f64,
    pub results: Vec<Result>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "snake_case")]
pub enum RejectReason {
    PositionLimit,
    NotOpen,
    NoLiquidity,
    BadTick,
    BadPrice,
    UnknownOrder,
    NotYourOrder,
    NotHost,
    UnknownPlayer,
}

/// Server -> client. Internally tagged on `t` to match the TS discriminated union.
#[derive(Debug, Clone, Serialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(tag = "t", rename_all = "camelCase")]
pub enum ServerEvent {
    #[serde(rename_all = "camelCase")]
    Snapshot {
        #[ts(type = "number")]
        seq: u64,
        session: SessionMeta,
        players: Vec<Player>,
        book: BookState,
        trades: Vec<Trade>,
        you: Option<YouState>,
        settlement: Option<Settlement>,
    },
    #[serde(rename_all = "camelCase")]
    PlayerJoined {
        #[ts(type = "number")]
        seq: u64,
        player: Player,
    },
    #[serde(rename_all = "camelCase")]
    PhaseChanged {
        #[ts(type = "number")]
        seq: u64,
        phase: Phase,
    },
    #[serde(rename_all = "camelCase")]
    OrderAdded {
        #[ts(type = "number")]
        seq: u64,
        order: Order,
    },
    #[serde(rename_all = "camelCase")]
    OrderCancelled {
        #[ts(type = "number")]
        seq: u64,
        order_id: String,
        player_id: String,
    },
    #[serde(rename_all = "camelCase")]
    Trade {
        #[ts(type = "number")]
        seq: u64,
        trade: Trade,
        resting_order_id: String,
    },
    #[serde(rename_all = "camelCase")]
    Settled {
        #[ts(type = "number")]
        seq: u64,
        true_value: f64,
        results: Vec<Result>,
    },
    #[serde(rename_all = "camelCase")]
    Rejected {
        reason: RejectReason,
        message: String,
    },
    #[serde(rename_all = "camelCase")]
    Error { message: String },
}

/// Client -> server.
#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(tag = "t", rename_all = "camelCase")]
pub enum ClientCommand {
    #[serde(rename_all = "camelCase")]
    PlaceOrder {
        side: Side,
        price: f64,
    },
    #[serde(rename_all = "camelCase")]
    CancelOrder {
        order_id: String,
    },
    #[serde(rename_all = "camelCase")]
    Take {
        direction: Direction,
    },
    #[serde(rename_all = "camelCase")]
    Resync {
        #[ts(type = "number")]
        from_seq: u64,
    },
    OpenTrading,
    CloseTrading,
    #[serde(rename_all = "camelCase")]
    Settle {
        true_value: f64,
    },
    /// Host-only. Adds one bot.
    ///
    /// Bots have no view on price. They buy and sell at random, at a rate the
    /// host sets — they are order flow, not opinion.
    AddBot,
    /// Host-only. How the bots behave.
    ///
    /// `orders_per_minute` is per bot. `buy_bias` is 0..1: 0.5 is even, higher
    /// buys more than it sells, so the host can lean the flow one way.
    #[serde(rename_all = "camelCase")]
    SetBotFlow {
        orders_per_minute: f64,
        buy_bias: f64,
    },
    /// Host-only. Stops the bots trading.
    RemoveBots,
}

/* -- HTTP payloads -------------------------------------------------------- */

#[derive(Debug, Deserialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub struct CreateSessionRequest {
    pub question: String,
    pub unit: String,
    pub tick_size: Option<f64>,
    #[ts(type = "number")]
    pub position_limit: i64,
}

#[derive(Debug, Serialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub struct CreateSessionResponse {
    pub code: String,
    pub host_token: String,
}

#[derive(Debug, Deserialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub struct JoinSessionRequest {
    pub name: String,
}

#[derive(Debug, Serialize, TS)]
#[ts(export, export_to = "../../web/src/generated/")]
#[serde(rename_all = "camelCase")]
pub struct JoinSessionResponse {
    pub player_id: String,
    pub player_token: String,
}
